stdlib/public/core/KeyPathIterable.swift

```diff
@@ -0,0 +1,112 @@
+//===-- KeyPathIterable.swift ---------------------------------*- swift -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the KeyPathIterable protocol.
+//
+//===----------------------------------------------------------------------===//
+
+//===----------------------------------------------------------------------===//
+// KeyPathIterable
+//===----------------------------------------------------------------------===//
+
+/// An implementation detail of `KeyPathIterable`; do not use this protocol
+/// directly.
+public protocol _KeyPathIterableBase {
+  var _allKeyPathsTypeErased: [AnyKeyPath] { get }
+  var _recursivelyAllKeyPathsTypeErased: [AnyKeyPath] { get }
+}
+
+/// A type whose values provides custom key paths to properties or elements.
+public protocol KeyPathIterable : _KeyPathIterableBase {
+  /// A type that can represent a collection of all key paths of this type.
+  associatedtype AllKeyPaths : Collection
+    where AllKeyPaths.Element == PartialKeyPath<Self>
+
+  /// A collection of all custom key paths of this value.
+  var allKeyPaths: AllKeyPaths { get }
+}
+
+public extension KeyPathIterable {
+  /// An array of all custom key paths of this value and any custom key paths
+  /// nested within each of what this value's key paths refers to.
+  var recursivelyAllKeyPaths: [PartialKeyPath<Self>] {
+    var result: [PartialKeyPath<Self>] = []
+    for kp in allKeyPaths {
+      result.append(kp)
+      if let nested = self[keyPath: kp] as? _KeyPathIterableBase {
+        for nkp in nested._recursivelyAllKeyPathsTypeErased {
+          result.append(kp.appending(path: nkp)!)
+        }
+      }
+    }
+    return result
+  }
+}
+
+public extension KeyPathIterable {
+  var _allKeyPathsTypeErased: [AnyKeyPath] {
+    return allKeyPaths.map { $0 as AnyKeyPath }
```

Comment:
Since key paths are classes, you can cast the array for these two instead of mapping:

```swift
return allKeyPaths as [AnyKeyPath]
```

(This is really cool functionality!)

Reply:
It's not really directly convertible. I'm getting this error:

```console
/Users/rxwei/Development/Swift/swift-tf-source/swift/stdlib/public/core/KeyPathIterable.swift:57:24: error: 'Self.AllKeyPaths' is not convertible to '[AnyKeyPath]'; did you mean to use 'as!' to force downcast?
    return allKeyPaths as [AnyKeyPath]
           ~~~~~~~~~~~~^~~~~~~~~~~~~~~
                       as!
```

Reply:
If that works on master, then it's probably because the changes haven't been merged. The tensorflow branch is a pretty old (2018-11-24).

Reply:
Ah, you’re right — I’d missed that `allKeyPaths` isn’t an array here, it’s an instance of the associated type. Carry on!

Reply:
Ah right. I forgot that too :)